### ISSUE TITLE
Include user parameters in painless reduce script context

### ISSUE THREAD

**Issue by nbrahms** [elastic/elasticsearch#42046]
Hello,

I'd like to request that script parameters be available in scripted metric reduce scripts.

Here's a toy example:
```
GET foo/_search
{
  "query" : {
    "match_all": {}
  },
  "size": 0,
  "aggregations": {
    "foo": {
      "scripted_metric": {
        "params": {
          "foo": 2
        },
        "init_script": "state.bla = 0",
        "map_script": "state.bla += 2",
        "combine_script": "return state.bla",
        "reduce_script": "int r = 0; for (s in states) r += s; return r * params.foo"
      }
    }
  }
}
```

which returns:
```
{
  "error": {
    "root_cause": [],
    "type": "search_phase_execution_exception",
    "reason": "",
    "phase": "fetch",
    "grouped": true,
    "failed_shards": [],
    "caused_by": {
      "type": "script_exception",
      "reason": "runtime error",
      "script_stack": [
        "return r * params.foo",
        "                 ^---- HERE"
      ],
      "script": "int r = 0; for (s in states) r += s; return r * params.foo",
      "lang": "painless",
      "caused_by": {
        "type": "null_pointer_exception",
        "reason": null
      }
    }
  },
  "status": 503
}
```

IMO it should be pretty common to parameterize reduction (in my case, I want to control a final sort order). Can we throw this on the roadmap?

**Comment by nbrahms** [comment]
For what it's worth, this is the current workaround I'm using:

```{
  "query" : {
    "match_all": {}
  },
  "size": 0,
  "aggregations": {
    "foo": {
      "scripted_metric": {
        "params": {
          "foo": 2
        },
        "init_script": "state.bla = 0; state.foo = params.foo",
        "map_script": "state.bla += 2",
        "combine_script": "return state",
        "reduce_script": "int r = 0; for (s in states) r += s.bla; return r * states[0].foo"
      }
    }
  }
}```

**Comment by jdconrad** [comment]
Seems like a bug.  Will look into this soon.

**Comment by nbrahms** [comment]
Awesome, thanks for the fast response!

**Comment by jdconrad** [comment]
This fix is now in master for 8.0 and will be backported to 7.3 as well.

**Comment by booster96** [comment]
how can I use some aggregated value(for ex some of any column) which is outside scripted_metric and then use that value in params.foo inside scripted_metric. Please advice
